src/collector.rs

```diff
@@ -0,0 +1,81 @@
+/// Epoch-based garbage collector.
+///
+/// # Examples
+///
+/// ```
+/// use crossbeam_epoch::Collector;
+///
+/// let collector = Collector::new();
+///
+/// let handle = collector.handle();
+/// drop(collector); // `handle` still works after dropping `collector`
+///
+/// handle.pin(|scope| {
+///     scope.flush();
+/// });
+/// ```
+
+use std::sync::Arc;
+use internal::{Global, Local};
+use scope::{Scope, unprotected};
+
+/// An epoch-based garbage collector.
+pub struct Collector(Arc<Global>);
+
+/// A handle to a garbage collector.
+pub struct Handle {
+    global: Arc<Global>,
+    local: Local,
+}
+
+impl Collector {
+    /// Creates a new collector.
+    pub fn new() -> Self {
+        Self { 0: Arc::new(Global::new()) }
+    }
+
+    /// Collect several bags from the global garbage queue and destroy their objects.
+    ///
+    /// # Safety
+    ///
+    /// It is assumed that no handles are concurrently accessing objects in the global garbage
+    /// queue. Otherwise, the behavior is undefined.
+    #[inline]
+    pub unsafe fn collect(&self) {
+        unprotected(|scope| self.0.collect(scope))
```

Comment:
Could you give an example scenario where this fails/crashes? Looking at the signature of `Global::collect`, it seem safe. The function collects garbage in accordance with the epoch invariant of two epochs having passed since a bag was added to the queue, so there is no way that any handle is concurrently accessing any of the objects that can be deleted (are old enough). The difference between this and calling `Global::collect` inside `pin` is that in the case of `pin`, `try_advance` will never go beyond the current epoch, while here it might, but that's fine, because the thread calling `Collector::collect` is not holding any pointers from the current epoch (unless `Collector::collect` is called in an `unprotected` scope, but that's the user's fault for using a non-exclusive pointer in `unprotected`) and if any other thread/handle has such pointers, it will be pinned.

Reply:
I think the unsafety comes from the fact that `collect()` `defer_drop(ptr)`s (or `defer(move || { ptr.into_owned()})`s in the recent terminology) a pointer `ptr` in the global garbage queue.

Suppose two threads are concurrently calling `collect()`, and a thread just deallocated a node in the garbage queue.  The node is immediately deallocated since we are calling `Global::collect()` within a `unprotected()` scope.  However, the other thread may try to read from a deallocated object, invoking undefined behavior.

I spent some time to come up with this explanation, which basically means it's a documentation bug.  I updated the function doc.  Thank you for pointing out this!

Reply:
Hah, that's pretty well-hidden unsafety! Thanks for explaining this, it's interesting to see how it propagates down the call stack. On the function comment, since the garbage queue is an implementation detail and probably shouldn't be mentioned in user-facing documentation, how about simply stating instead that:

> It is assumed that no other thread is concurrently using this collector or any of its handles. Otherwise, the behavior is undefined and might lead to memory corruption.

and then explaining the following in a comment inside the function body:

> This is unsafe, because the calling thread may deallocate a node in the global garbage queue immediately due to the use of `unprotected`,  and any other thread that calls `Global::collect()` can then read uninitialized memory while walking through the queue.

Reply:
So unsafety stems from the fact that we actually don't pin here (we have `unprotected` instead `pin`). And the reason why we don't pin is that there isn't a good way to. We could create a one-off handle to safely collect garbage:

```rust
self.handle().pin(|scope| self.0.collect(scope));
```

but allocating, registering and unregistering a handle just to collect some garbage is obviously not a great solution. I think the right thing to do here would be to implement a special kind of *anonymous pinning* - pinning without a handle (`Collector::pin`). The `try_advance` method would then also check whether anyone has been anonymously pinned.

Reply:
@stjepang I thought about your proposal.  I think the following `Collector::pin()` would be correct:

```rust
// Now a `Collector` has the field `anonymous_pin_counter: AtomicUsize`

fn pin<F, R>(&self, f: F) -> R where ... {
    fetch_add(self.anonymous_pin_counter, 1, Relaxed);
    fence(SeqCst);
    let scope = Scope {
        global: self.0,
        bag: Bag::new()
    };

    defer! {
        fetch_add(self.anonymous_pin_counter, -1, Relaxed);
    }

    let result = f(&scope);
    self.0.push_bag(&mut scope.bag, &scope);
    result
}

fn try_advance(...) {
  'L40: let G = EPOCH.load(Relaxed);
  'L41: atomic::fence(SeqCst);

  'LADD: if (anonymous_pin_counter.load(Relaxed) > 0) return;   // ADDED!
  'L42: for th in threads {
  'L43:   let e = th.epoch.load(Relaxed);
  'L44:   if e != SENTINEL && e != G { return; }
  'L45: }
  'L46: atomic::fence(Acquire);

  'L47: EPOCH.store(G + 1, Release);

  'L48: // now objects marked with `<= e - 1` can be deallocated.
}
```

for the [same reason as the handles](https://github.com/crossbeam-rs/rfcs/issues/16).  (Actually it was missing in my [previous proof of Crossbeam](https://github.com/crossbeam-rs/rfcs/blob/master/text/2017-07-23-relaxed-memory.md)!).  Now it's enough for `try_advance()` to check if `anonymous_pin_counter` is zero.


Though this is correct (at least in my opinion), it will block the epoch advance more than necessary.  Once a collector is anonymously pinned, the global epoch can't advance.  On the other hand, if we create a one-off handle, the global epoch can advance.  I cannot estimate the performance impact precisely, but I think it can quite bad at least for some use cases.  I'm not sure this "blocking" nature can easily be lifted.


So we have at least three candidate implementation ideas: (1) the unsafe version in the RFC, (2) safe "blocking" version presented above, and (3) hypothetical safe nonblocking version we will think of.  For this reason, I'd like to postpone the discussion on `Collector::collect()` to a later RFC.

Reply:
@jeehoonkang Yeah, that seems correct.

Regarding the blocking nature of this algorithm, maybe it's easy to fix after all... :)

Instead of having one counter, let's have two:
```rust
anon_count_even: AtomicUsize,
anon_count_odd: AtomicUsize,
```

To anonymously pin a thread, first load `EPOCH`. Then, if `epoch & 2 == 0` increment `anon_count_even`, otherwise increment `anon_count_odd`.

In `try_advance` just check the counter of opposite parity:
```rust
if G & 2 == 0 && anon_count_odd.load(Relaxed) > 0 { return }
if G & 2 == 1 && anon_count_even.load(Relaxed) > 0 { return }
```

Using this scheme the global epoch can still make progress even in the presence of anonymous pins. Proof of correctness might be a bit tricky, but I believe this should work. :)

Reply:
I like your idea :)  Very neat!

Reply:
... now I wonder if why we need to maintain the registry as a participant list.  @stjepang suggested that two `AtomicUsize` are enough for supporting anonymous handles.  If it's true, why don't we extend this idea further to normal handles?

Reply:
Because that would create too much contention. If you have a lot of threads modifying two atomic numbers, they will constantly invalidate each other's cache. I did some measurements with this method last year, and this contention would sometimes severely affect performance.

With a linked list every thread modifies its own atomic number so there is no cache invalidation.

Reply:
Beyond performance reasons, I doubt the correctness of the parity method (at least in C/C++ standard).  (For the same reason, I doubt the correctness of the proposed implementation of `Collector::pin()`).  Maybe we can adjust the proof eventually, but for now I couldn't manage to adjust the existing Crossbeam proof for the version with two atomic numbers..

The key reason is, if `pin()` reads a very old stale value of the global epoch, then the pinning thread may be able to read an object, while a concurrent thread is collecting the object.  The parity is not precise abstraction of the local epochs.